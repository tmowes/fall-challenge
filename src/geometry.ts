import { random } from "./random";
import { PriorityQueue } from "./queues";

export interface Point {
  x: number;
  y: number;
}

export function manhattanDistance(pointA: Point, pointB: Point): number {
  return Math.abs(pointA.x - pointB.x) + Math.abs(pointA.y - pointB.y);
}

export function areEqual(pointA: Point, pointB: Point): boolean {
  return pointA.x === pointB.x && pointA.y === pointB.y;
}

export function findRandomAvailablePosition(map: string[][]): Point {
  while (true) {
    let y = Math.floor(random() * map.length);
    let x = Math.floor(random() * map[0].length);

    if (map[y][x] === " ") {
      return { x, y };
    }
  }
}

// A* algorithm using manhattan distance
export function findPath(
  map: string[][],
  srcPoint: Point,
  dstPoint: Point
): { path: Point[]; distance: number } {
  type VisitingPoint = {
    point: Point;
    visited: boolean;
    distanceToSrc: number;
    throughPoint: VisitingPoint;
  };
  const pointsQueue = new PriorityQueue<VisitingPoint>({
    comparator: function (pointA, pointB) {
      const pointATotalDistance =
        pointA.distanceToSrc + manhattanDistance(dstPoint, pointA.point);
      const pointBTotalDistance =
        pointB.distanceToSrc + manhattanDistance(dstPoint, pointB.point);
      return pointATotalDistance - pointBTotalDistance;
    },
  });

  let visitedPointsMatrix: VisitingPoint[][] = [];

  for (let y = 0; y < map.length; y++) {
    visitedPointsMatrix[y] = [];
    for (let x = 0; x < map[y].length; x++) {
      visitedPointsMatrix[y][x] =
        map[y][x] === "#"
          ? null
          : {
              point: { x, y },
              visited: false,
              distanceToSrc: Number.POSITIVE_INFINITY,
              throughPoint: null,
            };
    }
  }

  visitedPointsMatrix[srcPoint.y][srcPoint.x].distanceToSrc = 0;
  pointsQueue.queue(visitedPointsMatrix[srcPoint.y][srcPoint.x]);

  while (pointsQueue.length > 0) {
    const currentPointToVisit = pointsQueue.dequeue();
    currentPointToVisit.visited = true;

    if (areEqual(currentPointToVisit.point, dstPoint)) {
      let path: VisitingPoint[] = [currentPointToVisit];
      let currentPathPoint = currentPointToVisit;
      while (currentPathPoint.throughPoint) {
        path = [currentPathPoint.throughPoint, ...path];
        currentPathPoint = currentPathPoint.throughPoint;
      }

      return {
        path: path.map((visitingPoint) => visitingPoint.point),
        distance: path[path.length - 1].distanceToSrc,
      };
    }

    const width = visitedPointsMatrix[currentPointToVisit.point.y].length;

    const northPoint =
      visitedPointsMatrix[currentPointToVisit.point.y - 1] &&
      visitedPointsMatrix[currentPointToVisit.point.y - 1][
        currentPointToVisit.point.x
      ];
    const southPoint =
      visitedPointsMatrix[currentPointToVisit.point.y + 1] &&
      visitedPointsMatrix[currentPointToVisit.point.y + 1][
        currentPointToVisit.point.x
      ];
    const eastPoint =
      visitedPointsMatrix[currentPointToVisit.point.y][
        (currentPointToVisit.point.x + 1) % width
      ];
    const westPoint =
      visitedPointsMatrix[currentPointToVisit.point.y][
        currentPointToVisit.point.x -
          1 +
          (currentPointToVisit.point.x - 1 < 0 ? width : 0)
      ];
    if (northPoint) {
      if (northPoint.distanceToSrc > currentPointToVisit.distanceToSrc + 1) {
        northPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        northPoint.throughPoint = currentPointToVisit;
      }
      if (!northPoint.visited) {
        pointsQueue.queue(northPoint);
      }
    }
    if (southPoint) {
      if (southPoint.distanceToSrc > currentPointToVisit.distanceToSrc + 1) {
        southPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        southPoint.throughPoint = currentPointToVisit;
      }
      if (!southPoint.visited) {
        pointsQueue.queue(southPoint);
      }
    }
    if (eastPoint) {
      if (eastPoint.distanceToSrc > currentPointToVisit.distanceToSrc + 1) {
        eastPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        eastPoint.throughPoint = currentPointToVisit;
      }
      if (!eastPoint.visited) {
        pointsQueue.queue(eastPoint);
      }
    }
    if (westPoint) {
      if (westPoint.distanceToSrc > currentPointToVisit.distanceToSrc + 1) {
        westPoint.distanceToSrc = currentPointToVisit.distanceToSrc + 1;
        westPoint.throughPoint = currentPointToVisit;
      }
      if (!westPoint.visited) {
        pointsQueue.queue(westPoint);
      }
    }
  }

  return null;
}
