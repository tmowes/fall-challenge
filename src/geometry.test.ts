import { findPath } from "./geometry";

it(`should get shortest past with A*`, () => {
  expect(findPath(getMap(), { x: 0, y: 0 }, { x: 4, y: 3 })).toMatchSnapshot();
});

it(`should get shortest past with A* with pac in the way`, () => {
  let map = getMap();
  map[1][3] = "#";
  expect(findPath(map, { x: 0, y: 0 }, { x: 4, y: 3 })).toMatchSnapshot();
});

function getMap(): string[][] {
  return [
    [" ", " ", " ", " ", " "],
    [" ", "#", "#", " ", "#"],
    [" ", "#", "#", " ", "#"],
    [" ", " ", "#", " ", " "],
    ["#", " ", " ", " ", " "],
  ];
}
