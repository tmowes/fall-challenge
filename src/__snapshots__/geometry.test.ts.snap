// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get shortest past with A* 1`] = `
Object {
  "distance": 6,
  "path": Array [
    Object {
      "x": 0,
      "y": 0,
    },
    Object {
      "x": 4,
      "y": 0,
    },
    Object {
      "x": 3,
      "y": 0,
    },
    Object {
      "x": 3,
      "y": 1,
    },
    Object {
      "x": 3,
      "y": 2,
    },
    Object {
      "x": 3,
      "y": 3,
    },
    Object {
      "x": 4,
      "y": 3,
    },
  ],
}
`;

exports[`should get shortest past with A* with pac in the way 1`] = `
Object {
  "distance": 4,
  "path": Array [
    Object {
      "x": 0,
      "y": 0,
    },
    Object {
      "x": 0,
      "y": 1,
    },
    Object {
      "x": 0,
      "y": 2,
    },
    Object {
      "x": 0,
      "y": 3,
    },
    Object {
      "x": 4,
      "y": 3,
    },
  ],
}
`;
